### PR TITLE
Add powmod to BigInt

### PR DESCRIPTION
**Pull request by dlesnoff**
Following request of Issue #45, I have implemented modular exponentiation following SageMath implementation. The only discrepancy being it does not yet accept negative exponent.
I have added some tests too. Some checks uses mathematical properties as the little fermat's theorem. We might want to reduce them in the future, as they take a little time.
For negative exponentiation, we need to invert first using an extended euclidean algorithm.
Some examples did not compile anymore, due to a change to the backwards index operator .. < needs to be written as ..< .

### CODE REVIEW

File: src/bigints.nim [hunk 1]
```diff
@@ -959,3 +959,33 @@ iterator `..<`*(a, b: BigInt): BigInt =
   while res < b:
     yield res
     inc res
+
+func powmod*(base, exponent, modulus: BigInt): BigInt =
+  ## Compute modular exponentation of `base` with power `exponent` modulo `modulus`.
+  if modulus == 0:
+    raise newException(DivByZeroDefect, "modulus must be nonzero.")
```
Comment (konsumlamm) [inline src/bigints.nim:966]:
```suggestion
    raise newException(DivByZeroDefect, "modulus must be nonzero")
```
Current error messages don't end with `.` either (besides, this is not a full sentence anyway).

File: src/bigints.nim [hunk 2]
```diff
@@ -959,3 +959,33 @@ iterator `..<`*(a, b: BigInt): BigInt =
...
+  if modulus == 1:
+    return zero
+  if exponent < 0:
+    # let baseInv = inverse_mod(base, modulus)
+    # return powmod(baseInv, -exponent, modulus)
+    raise newException(ValueError, "Not yet Implemented!")
+  if exponent == 0:
+    if base == 0:
+      raise newException(ValueError, "0^0 is undefined.")
```
Comment (konsumlamm) [inline src/bigints.nim:975]: For `pow` and ``math.`^` ``, we have 0^0 = 1, which is also common in practice (although of course it's technically indeterminate). I'm not sure how common this is in modular arithmetic though.

File: src/bigints.nim [hunk 3]
```diff
@@ -959,3 +959,33 @@ iterator `..<`*(a, b: BigInt): BigInt =
...
+    return one
+
+  var
+    exp = exponent
+    basePow = ((base mod modulus) + modulus) mod modulus # Base stays in [0, m-1]
```
Comment (konsumlamm) [inline src/bigints.nim:980]: So the final result will always be in [0, m-1]? What happens if `modulus` is negative? This should be documented.

File: src/bigints.nim [hunk 4]
```diff
@@ -959,3 +959,33 @@ iterator `..<`*(a, b: BigInt): BigInt =
...
+  while (exp and one) == zero:
+    basePow = basePow * basePow mod modulus
+    exp = exp shr 1
+  result = basePow
+  exp = exp shr 1
```
Comment (konsumlamm) [inline src/bigints.nim:985]: Why is this part needed? Is it just an optimization?

File: src/bigints.nim [hunk 5]
```diff
@@ -959,3 +959,33 @@ iterator `..<`*(a, b: BigInt): BigInt =
...
+  while exp != zero:
+    basePow = (basePow*basePow) mod modulus
+    if (exp and one) != zero:
+      result = (result*basePow) mod modulus
```
Comment (konsumlamm) [inline src/bigints.nim:989]:
```suggestion
    basePow = (basePow * basePow) mod modulus
    if (exp and one) != zero:
      result = (result * basePow) mod modulus
```

File: tests/tbigints.nim [hunk 6]
```diff
@@ -379,5 +379,37 @@ template main() =
     doAssert pow(zero, 0) == one
     doAssert pow(zero, 1) == zero
 
+  block: #pow_mod
```
Comment (konsumlamm) [inline tests/tbigints.nim:382]:
```suggestion
  block: # powmod
```

File: tests/tbigints.nim [hunk 7]
```diff
@@ -379,5 +379,37 @@ template main() =
...
+    let a = "30292868".initBigInt
+    let p = "60449131".initBigInt # p is prime
+    let two = 2.initBigInt
+    doAssert powmod(a, two, p) == "25760702".initBigInt
+    # Fermat's little theorem: a^p ≡ a, for all a in F_p
```
Comment (konsumlamm) [inline tests/tbigints.nim:387]:
```suggestion
    # Fermat's little theorem: a^p ≡ a mod p
```
I think this is clearer for people that don't know what `F_p` is.

File: tests/tbigints.nim [hunk 8]
```diff
@@ -379,5 +379,37 @@ template main() =
...
+    doAssert powmod(a, p, p) == a
+    # Euler's identity a^(p-1) \equiv 1 \bmod p
+    doAssert powmod(a, p - one, p) == one
+    # We can invert a using euler's identity / fermat's little theorem
```
Comment (konsumlamm) [inline tests/tbigints.nim:391]:
```suggestion
    # We can invert a using Euler's identity / Fermat's little theorem
```

File: tests/tbigints.nim [hunk 9]
```diff
@@ -379,5 +379,37 @@ template main() =
...
+    doAssert powmod(a, p - two, p) == "51713091".initBigInt
+    # We can reduce the exponent modulo phi(p) = p - 1, since p is prime
+    doAssert powmod(a, 2.initBigInt*p, p) == (a*a mod p)
+
+    let p2 = 761.initBigInt
+    var a2 = 1.initBigInt
+    # Fermat's little theorem: a^p ≡ a, for all a in F_p
```
Comment (konsumlamm) [inline tests/tbigints.nim:398]:
```suggestion
    # Fermat's little theorem: a^p ≡ a mod p
```
See above.

**Comment by konsumlamm** [comment]
Can you rebase on the latest master? That also makes the diff easier to read.

File: src/bigints.nim (hunk 2 above)
Comment (dlesnoff) [reply src/bigints.nim:975]: If this is true for other functions, then it looks like it would be nitpicking to raise a ValueError for this special case. I have removed it.

File: src/bigints.nim (hunk 3 above)
Comment (dlesnoff) [reply src/bigints.nim:980]: The final result will always be in [0, m-1] contrary to the `mod` keyword for integers.
It does not make sense for the modulus to be negative, I will add an exception.
I have not tested what happens if the base is negative. I will add some tests for this case.

File: src/bigints.nim (hunk 4 above)
Comment (dlesnoff) [reply src/bigints.nim:985]: It does not work without, I am not sure why, I will look into it tomorrow.

**Comment by dlesnoff** [comment]
I have deleted my fork and started from anew, as I do not know how to rebase upstream repositories. I will probably have to make another pull request.

File: src/bigints.nim (hunk 4 above)
Comment (konsumlamm) [reply src/bigints.nim:985]: I think it should work if you replace this by `result = one`. Then we shouldn't need an extra case for `exponent == 0` either.
